Parallel process image data

HughMacdonald opened the pull request:
This PR adds ImageAlgo methods for parallel processing of tiles/channels, and re-implements both ImagePlug::image() and ImageWriter to use this.

ImageAlgo has 4 new methods:

* parallelProcessTiles
  * Two versions
    * One takes an imagePlug, a functor and an optional window to process (default will be imagePlug->dataWindowPlug()
    * One takes the same parameters as above, plus a vector<string> of channel names, and will call the functor once per tile per channel
  * This will call the defined functor once per tile (per channel)
* parallelGatherTiles
  * Same two versions as above
  * Also takes:
    * Gather functor, which will be called for the same tiles/channels, but in series.
    * tileOrder, which defines the order that the gather functor is called in. One of:
      * Unordered : No guarantee of ordering. Will tend towards bottom to top.
      * BottomToTop : Guaranteed iteration through channels, tiles in X, tiles in Y, from the bottom
      * TopToBottom : Same as BottomToTop, but from the top in Y

ImagePlug::image(), which previously dealt with its own parallelising, now uses parallelProcessTiles() to set off parallel processing of channel data to create an ImagePrimitive.
The parallelising that it used to do had a bug where it would often end up processing the same tiles multiple times, due to a misunderstanding of how tbb's blocked_range dealt with splitting up ranges. It would sometimes pass through ranges that were smaller than `grainsize`. As such, it now does all of the splitting using IDs, and then these are turned back into coordinates internally.

ImageWriter used to call ImagePlug::image() to get a full image before writing it. It now has its own classes for storing the data and writing it to an OIIO::ImageOutput object. 


There is also a new test - ImageNodeTest.testCacheThreadSafety(), which uses GafferImageTest.traverseTiles() to call imagePlug->channelData() on each tile and channel.

HughMacdonald commented:
Hmmm - Travis is properly crashing out on ImageWriterTest.testWriteEmptyImage

That's not a crash I've seen myself locally (OSX, debug build). Going to try a debug build on Linux and a release build on OSX and see if I can get it to crash.

Move along people. Nothing to see here.

HughMacdonald commented:
Comments so far addressed in new commit.

HughMacdonald commented:
Also, I've been able to replicate the crash on a Linux machine. I'm going to get my Ubuntu VM back up and running on my OSX machine, and I'll work on it there.

johnhaddon commented:
OK, I've done a partial review of the ImageWriter mostly just on the basic organisation - I'll take a closer look at the internal logic when you have the crash fixed.

Overally I'm feeling good about this though - it's nice to see the details of the threading getting abstracted away like this, and the performance improvements you've made are very welcome too :)

HughMacdonald commented:
Great - thanks.

I'm almost done making your suggested changes, then I'll have the `ImagePlug::imageHash()` re-implementation to do, and the crash to fix. I'll let you know when that's all in there.

Cheers

HughMacdonald commented:
I fixed the crash. After finding out what it was, I'm somewhat shocked that it wasn't crashing on OSX!

I'm just working on ImagePlug::imageHash() at the moment. Would I be right in saying that it does make a difference in which order hashes are appended?

johnhaddon commented:
Good news! And yes, order is important. Don't worry about matching the current imageHash() results exactly if that would be fiddly, but do ensure that the order of appending is consistent in the new version, and have a unit test assert that repeated calls give the same result...

HughMacdonald commented:
Cool.

I've just pushed a couple of commits that adds the imageHash() parallel processing, so all should be good, as long as Mr Travis agrees.

HughMacdonald commented:
Just done a small tweak to the test to get it to run imageHash() 20 times in a row and check if they're all the same, just in case parallelisation was causing it to run the same order some times. It should be good now.

HughMacdonald commented:
That looks like quite a dramatic crash there! Will have to dig into that one a bit... It might be a few days until I can get into this.

HughMacdonald commented:
I think I've fixed the crash.

Note to self - don't trust OSX to protect from crashes. Always run tests on Linux before submitting a PR...

johnhaddon commented:
Am I right in thinking that the ImageWriter is trying to deal with situations whereby the output tile size is not exactly the same as Gaffer's native tile size?

HughMacdonald commented:
Yes. It was written in a way that was general enough that it could deal with these situations. I don't believe it adds any significant overhead vs assuming the same tile size as Gaffer's internal tile size.

However, output tiles, even if the same size, will usually not line up to Gaffer tiles, as output tiles align to the data window origin, and are based on the top left rather than the bottom left.

andrewkaufman commented:
Is this expected to be fully functional currently? I'm getting a seg fault with a fairly awkward use case (a custom node derived from `ImagePrimitiveSource`) which should bottleneck the threading, but is instead crashing immediately. I haven't looked into it further since I didn't know where we're at on thinking there are outstanding crashes here or not.

HughMacdonald commented:
I was thinking it was fully-functional, so would be good to know how best to replicate this crash...

andrewkaufman commented:
I'll try to get you a minimal repro. Its not actually a use case we want in the long run (ImagePrimitiveSource should die), but until we have a proper Warp base node in Gaffer, I'm stuck diving out into IECore::ModifyOps unfortunately.

HughMacdonald commented:
Cool - let me know when you have something and I'll figure out what's happening.

Cheers

johnhaddon commented:
@HughMacdonald, my question about arbitrary tile sizes was motivated mainly by the complexity of trying to review and maintain the code, rather than any runtime overhead. I have to admit to glazing over when trying to follow all the logic in ImageWriter, and I had hoped that things could be simplified somewhat. But as you say the tiles won't line up anyway even if of the same size, so I guess some of the complexity is needed.

@andrewkaufman, it might be good if you could `git bisect` to find the commit that's causing the crash. I had been considering merging the first half of this PR on it's own and then continuing to improve the ImageWriter part - if the problem is in the ImageWriter then I'll be even more motivated to do so.

Just to confuse things further, I did some tidying up as I reviewed this branch, which can be found here :

https://github.com/johnhaddon/gaffer/commits/HughMacdonald-parallelProcessImageData

@HughMacdonald, I think it's worth paying particular attention to https://github.com/johnhaddon/gaffer/commit/51ee62c46fd4a7930a457fc4014481f5af227774. Assuming I've understood right, there really wasn't any need for that intermediate storage and indexing logic I removed - this is the kind of simplification we're striving to enable when introducing abstractions like `parallelGatherTiles()`. I feel like there might be similar abstractions we could use to improve the readability of things like ImageWriter. Maybe some abstractions in ImageAlgo like "copy the data from this tile into that tile where they intersect"? No need to jump straight into that, but I think it might be worth bearing in mind for the future...

HughMacdonald commented:
> @HughMacdonald, my question about arbitrary tile sizes was motivated mainly by the complexity of trying to review and maintain the code, rather than any runtime overhead. I have to admit to glazing over when trying to follow all the logic in ImageWriter, and I had hoped that things could be simplified somewhat. But as you say the tiles won't line up anyway even if of the same size, so I guess some of the complexity is needed.

The only change to assuming that tiles are the same size would be that you would know that you'd never need more than 4 input tiles for an output tile, but I don't think that the code complexity would decrease at all for this.

I can talk you through the code tomorrow - it was a little tricky to figure out, and I get that it's a bit confusing, and I'd love to have a chat about how it could be simplified.

> @HughMacdonald, I think it's worth paying particular attention to johnhaddon@51ee62c.

I definitely prefer what you've done there. I'm not sure why I didn't go with parallelGatherTiles in the first place, as that's exactly what that was designed for. Mindfart from me on that one!

HughMacdonald commented:
I've been looking at what a generic "copyDataWindow" method in ImageAlgo would be.

It's tricky because it would also need to deal with interleaved channel data (which ImageWriter requires), and would we want it to be able to deal with input interleaved channel data too? What should it do if both the input and output data is interleaved?

This prototype would be able to do what ImageWriter requires, while also supporting the same thing on both input and output data. If input and output data have interleaved channels, it would still only copy 1 channel's worth of data.

    void copyDataWindow(
		ConstFloatVectorDataPtr inData,
		FloatVectorDataPtr outData,
		const Imath::Box2i &inWindow,
		const Imath::Box2i &outWindow,
		const bool inDataYDown = false,
		const bool outDataYDown = false,
		const size_t inDataPixelOffset = 0,
		const size_t outDataPixelOffset = 0,
		const size_t inDataPixelStride = 1,
		const size_t outDataPixelStride = 1 );

johnhaddon commented:
@HughMacdonald and I went over the ImageWriter part of this in person yesterday. Hugh spotted a couple of bugs and explained the general workings, and we discussed ways of improving the clarity of the code and reducing the likelihood of other bugs. The plan is to make those revisions and then have @andrewkaufman do a review with fresh eyes. Andrew, I don't suppose you've had any luck reproducing your crash or using `git bisect`?

andrewkaufman commented:
I haven't yet, no. I need to get my ImageReader PR ready for re-review, then I'll take a look at getting a repro here.

andrewkaufman commented:
```
25b6d2b90ead99f1e40c3772900d5602d75fa5aa is the first bad commit
commit 25b6d2b90ead99f1e40c3772900d5602d75fa5aa
Author: Hugh Macdonald <hugh.macdonald@nvizible.com>
Date:   Fri Oct 23 16:44:49 2015 +0200

    ImagePlug : Modify imageHash() to use parallelProcessTiles
```

Now that I actually read the commit names, that was kind of obvious since my node calls `imageHash()` an un-godly number of times...

andrewkaufman commented:
Ok, I have a repro. cherry-pick this one https://github.com/andrewkaufman/gaffer/commit/c81325dfe50e4c026f789ae39a86ca18465e26e2 and run `gaffer test GafferImageTest.CortexPremultTest`, which should seg fault.

If you then rebase to drop https://github.com/HughMacdonald/gaffer/commit/1c218c9dbd36523a5232740f48343272b9f650b1, the test passes.

johnhaddon commented:
Might be worth dropping in my simplified parallel imageHash() to see if that improves things. @HughMacdonald, I'm going to leave this to you for now so I can get on with other priorities, but I should be able to spend some time on it later next week if you don't get anywhere.

HughMacdonald commented:
I'll merge John's changes into my branch and test from there. Thanks guys

HughMacdonald commented:
I'm having trouble recreating this crash... Even compiling directly from https://github.com/andrewkaufman/gaffer/commit/c81325dfe50e4c026f789ae39a86ca18465e26e2 with all preceding commits, the test passes.

I've got two branches on mine - https://github.com/HughMacdonald/gaffer/tree/parallelProcessImageData (as before), which now contains @johnhaddon's commits (including the WIP one!), and https://github.com/HughMacdonald/gaffer/tree/parallelProcessImageData_crash, which should fail (but doesn't), which has @andrewkaufman's commit on the end.

HughMacdonald commented:
It's not crashing on OSX or Linux for me...

John significantly simplified the parallel imageHash, so please do try your end based on the latest commit in my parallelProcessImageData and see if it still happens...

andrewkaufman commented:
Its weird that you can't reproduce it, when it definitely crashed for me every time... but the new code seems to work fine for me now, so lets just ignore it and move on...

HughMacdonald commented:
Weird. Glad to hear it though. I'll work from the current head then... @johnhaddon, this does include your WIP commit...

HughMacdonald commented:
Hi all,

I've done a major update to this PR, so it should be good to review again. I've addressed @johnhaddon's notes, and added better comments, so it would be good, @andrewkaufman, if you could see if the methodology for tiles and scanline writing makes sense (the big comment blocks at the top of FlatTileWriter and FlatScanlineWriter)

Cheers

HughMacdonald commented:
Oh, and I also realised there was a place where we may have ended up writing image data that was in tiles, but outside the data window. I wrote a test that failed, and then fixed the issue.

andrewkaufman commented on src/GafferImage/ImageWriter.cpp:
```diff
@@ -66,13 +70,438 @@ static InternedString g_compressionQualityPlugName( "compressionQuality" );
+					writeTile( tileOrigin, m_tilesData[tileIndex] );
+					m_tilesData[tileIndex].reset();
+				}
+				else
```
Any reason this isn't an `else if` directly? I found the internal if/break logic confusing enough that I went back up and re-read the whole explanation comment to make sure I understood what was happening.

HughMacdonald replied:
Nope - it should be an else if there.

I'll add a comment to this method directly to explain what it's doing.

Hugh Macdonald
Nvizible - Visual Effects

020 3167 3860
07773 764708
www.nvizible.com

On 23 Nov 2015, at 19:26, Andrew Kaufman <notifications@github.com> wrote:

In src/GafferImage/ImageWriter.cpp
<https://github.com/ImageEngine/gaffer/pull/1517#discussion_r45648153>:

> +			}
> +		}
> +
> +		void writeFilledTiles()
> +		{
> +			size_t tileIndex;
> +			for( tileIndex = m_nextTileIndex; tileIndex < m_tilesData.size(); ++tileIndex )
> +			{
> +				Imath::V2i tileOrigin = outTileOrigin( tileIndex );
> +
> +				if( m_tilesFilled[tileIndex] )
> +				{
> +					writeTile( tileOrigin, m_tilesData[tileIndex] );
> +					m_tilesData[tileIndex].reset();
> +				}
> +				else

Any reason this isn't an else if directly? I found the internal if/break
logic confusing enough that I went back up and re-read the whole
explanation comment to make sure I understood what was happening.

—
Reply to this email directly or view it on GitHub
<https://github.com/ImageEngine/gaffer/pull/1517/files#r45648153>.

andrewkaufman commented on src/GafferImage/ImageWriter.cpp:
```diff
@@ -710,28 +1065,38 @@ void ImageWriter::execute() const
-	ImageSpec spec = createImageSpec( this, out.get(), dataWindow, displayWindow );
+	if( !empty( dataWindow ) )
+	{
+		exrDataWindow = imageFormat.toEXRSpace( dataWindow );
```
Can we make Format::toEXRSpace gracefully pass through empty windows? Or is there a downside to that?

andrewkaufman replied:
Oh wait, this isn't a pass though, is it. Its taking an un-initialized window and replacing it with a (0,0,->(0,0) window.

andrewkaufman commented:
I went through it, and I think I understand the whole Scanline vs Tile Writer logic. I made a couple fairly innocuous comments... overall I understand the reasoning for the approach, though it did take multiple read throughs of the big comment explanation.

I do find the code a bit difficult to navigate. Unfortunately I don't have any suggestions for improvement there as the topic seems complex enough that I'm not sure a simpler implementation is possible... Some of the maths I'm taking your word on, so it's entirely possible an off-by-one sort of error has gone unnoticed somewhere, but hopefully the test cases show that it hasn't.

What do you reckon @johnhaddon? What might be nice is some profile data, just to show what sort of speed vs memory tradeoff can be expected from this?

HughMacdonald commented:
I've addressed @andrewkaufman's note about writeFilledTiles

HughMacdonald commented:
I've just pushed some more changes, which splits parts of ImageAlgo out into BufferAlgo. Any methods that work on `Box2i`s are now in BufferAlgo. ImageAlgo contains the colour channel methods, and the parallel processing methods.

I'm thinking that copyBufferArea (currently here: https://github.com/HughMacdonald/gaffer/commit/6fc468a8fe87ad9e499165cd6b52c9b161873e78#diff-7d9752bff586c1eda65c502bb6202d2cR76) would make sense to move into BufferAlgo too.

johnhaddon commented:
Hi Hugh,
The build seems to be failing - I've got a commit that appears to fix it at my end :

https://github.com/johnhaddon/gaffer/commit/c6e24edb80a97d16e5163cd3ba0975e7efd59221

With that compilation is OK, but I get 4 failing ImageWriter tests. I don't know if that is because my fix is wrong (I don't _think_ it is though), or if there is some other problem. Could you cherry pick that commit onto your branch and see what you get?

Cheers...
John

HughMacdonald commented:
Thanks for that - I've been a little snowed under here so hadn't had a chance to get onto this. Hopefully will be back on it later this week and will let you know.

johnhaddon commented:
No worries - I've been slow to get to it in any case. I'm starting to look at an improved ImageView using the parallelGatherTiles stuff, so I may just cherry pick the earlier commits to get that moving.

HughMacdonald commented:
You know what - give me an hour or so this evening and I'll see if it's a quick fix. I've just put a render on, so should have a bit of time.

HughMacdonald commented:
Fixed it.

I was trying to be clever and not bother blanking out the scanlines before adding data. I added a fix because it had the possibility of writing data outside the data window, and this ended up meaning that old scanline data from a previous row of tiles could sneak through, which was what was happening here.

I've added back in code to blank out m_scanlinesData before each row of tiles.

HughMacdonald commented:
I found another compilation error. Fixed that one too. Let's see what Travis says!

HughMacdonald commented:
Hmm - so something very wrong there! I'll take a look first thing in the morning.

HughMacdonald commented:
It looks like I've fixed the crash...

HughMacdonald commented:
I've fixed the issue - it now passes your test, plus another few I added in.

johnhaddon commented:
I've put in a [new PR](#1586) which should make all that `BufferAlgo::empty()` hoop jumping unnecessary, so we can go back to the original implementation. I'd like to keep the additional tests you added, but once #1586 is merged, do you think you could rebase without c75b8f2 and 18ec477 Hugh?

HughMacdonald commented:
Our messages on the two PRs passed like ships in the night... Will do absolutely that. When do you think #1586 will be merged? It going to wait for @andrewkaufman? The reason I ask is because I'm off for a few days from tomorrow lunchtime...

johnhaddon commented:
I'd like @andrewkaufman to give #1586 a once over before merging, but I'm hoping he'll have time for that today (he recently got back from holidays, so I expect he's a bit slammed). I also just realised that ab9ea78 can be dropped as well.

HughMacdonald commented:
Cool stuff. I've dropped those commits, and will rebase on the new master when it exists.

johnhaddon commented:
Thought I'd do a bit of performance testing with the new ImageWriter while we're waiting on the other bits and bobs. With a contrived test upsizing from HD to 10K and writing the result I got a reduction in runtime of 65% and a reduction in peak memory usage of 60%. Nice one Hugh!

HughMacdonald commented:
Woop! Nice to know it was worth it!

HughMacdonald commented:
I've rebased on top of the new build changes, but I'm still seeing a failure on one of them so far (the other 3 are still running)

gcc 4.8 with C++98 is failing with the same error as before:

    include/GafferImage/BufferAlgo.inl: In member function 'void* tbb::interface6::internal::concrete_filter<T, void, Body>::operator()(void*) [with T = boost::tuples::tuple<long unsigned int, Imath_2_2::Vec2<int>, boost::intrusive_ptr<const IECore::TypedData<std::vector<float> > >, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type>; Body = GafferImage::Detail::GatherFunctorFilter<{anonymous}::FlatScanlineWriter, {anonymous}::TileProcessor>]':
    include/GafferImage/BufferAlgo.inl:51:3: error: assuming signed overflow does not occur when assuming that (X + c) < X is always false [-Werror=strict-overflow]
       if( window.max[i] <= window.min[i] )
       ^
    include/GafferImage/BufferAlgo.inl:51:3: error: assuming signed overflow does not occur when assuming that (X + c) < X is always false [-Werror=strict-overflow]
    cc1plus: all warnings being treated as errors
    scons: *** [src/GafferImage/ImageWriter.os] Error 1
    scons: building terminated because of errors.

HughMacdonald commented:
Same failure (but double the time!) on gcc 4.8 with C++11

HughMacdonald commented:
Plus, the default gcc, with C++98 flags the warning, but doesn't fail on it:

    In file included from ./build/gaffer-0.19.0.0-linux/include/tbb/tbb.h:67:0,
                     from include/GafferImage/ImageAlgo.inl:40,
                     from include/GafferImage/ImageAlgo.h:130,
                     from src/GafferImage/ImageWriter.cpp:56:
    include/GafferImage/BufferAlgo.inl: In member function 'void* tbb::interface6::internal::concrete_filter<T, void, Body>::operator()(void*) [with T = boost::tuples::tuple<long unsigned int, Imath_2_2::Vec2<int>, boost::intrusive_ptr<const IECore::TypedData<std::vector<float> > >, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type, boost::tuples::null_type>, Body = GafferImage::Detail::GatherFunctorFilter<{anonymous}::FlatScanlineWriter, {anonymous}::TileProcessor>]':
    include/GafferImage/BufferAlgo.inl:51:3: warning: assuming signed overflow does not occur when assuming that (X + c) < X is always false [-Wstrict-overflow]
    include/GafferImage/BufferAlgo.inl:51:3: warning: assuming signed overflow does not occur when assuming that (X + c) < X is always false [-Wstrict-overflow]

johnhaddon commented:
Yeah, I'd downgraded the warning only for GCC < 4.8, as in my testing it hadn't failed in 4.8. I'll just downgrade the warning for all versions and kick your tests off again.

HughMacdonald commented:
Cool. If you've got a commit you want me to cherry-pick into this PR (to keep the PR history), I can do that, or rebase on top of a new master. Let me know.

johnhaddon commented:
I'll fix things on master, which I think means you shouldn't need to do anything, because Travis should run the tests on a merge between master and your branch. Or so I think...

HughMacdonald commented:
Ahh - cool - I did not know that! I guess we shall see...

HughMacdonald commented:
Taking a quick look at the Travis docs, it says:

> Rather than test the commits from the branches the pull request is sent from, we test the merge between the origin and the upstream branch.

So this one should hopefully all work

johnhaddon commented:
Indeed it has...thanks Hugh.

HughMacdonald commented:
Yay! Thanks! Glad it's finally made it in!

Now to rebase deepImageCore (going to have a LOT of conflicts to resolve) and get back to that one!
